### DISCUSSION TITLE
RBG colour map and real values?

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/neper#466] [Q&A] [answered]
Hi! It's been a while and I missed it here!

Does anyone know how the colour map RGB values are translated from the real values? For example, how is a crystal orientation, say (37, 81, 112) in Euler convention translated to RGB values, if the default 'viridis' colour map is used?

**Comment by rquey** [comment, marked as answer]
Welcome back @xshang93!

It's described in, e.g., [-datacellcolscheme](https://neper.info/doc/neper_v.html#cmdoption-data-cell-poly-face-edge-ver-seed-crystal-vox-colscheme):

- For real values, the maps are described here: https://neper.info/doc/exprskeys.html#color-maps-for-real-values, and they are `viridis`, etc.
- For orientations, the typical maps are `rodrigues` and `ipf`, and they are shown in [this tutorial](https://neper.info/tutorials/orientation_color_key.html#rodrigues-color-key). True that they are not described in detail in the documentation. For `rodrigues`, the color is defined as {R,G,B}=255 x (r_i + sqrt(2) - 1) / (2 * (sqrt(2) - 1)), for cubic crystal symmetry.

I'll complete the documentation.

**Reply by xshang93** [reply]
Thanks a lot @rquey as always! I like it a lot the new tutorial sessions!

**Reply by xshang93** [reply]
Hey @rquey, just a quick follow up question, is there a quick formula of how the RGB values are calculated for ipf?

**Reply by xshang93** [reply]
Also, for `rodrigues`, if the colour is defined as {R,G,B}=255 x (r_i + sqrt(2) - 1) / (2 * (sqrt(2) - 1)), how is it made sure that the RGB values are always within (0,255)? I tried to make some calculation from the rodrigues values from my .tess file, but some of them give me very large numbers, and some gives me negative values. 
e.g. if an Euler angle is (4.4173*degree,132.5983*degree,171.6022*degree), the Rodrigues vector would be (7.32006,-65.1822,28.7767), which is (2380.74,-19936.7,8985.4) in RGB when using the formula above.

**Reply by mkasemer** [reply]
@xshang93: You have to return your Rodrigues vector to the fundamental region for the crystal system considered. You will then not have such a large Rodrigues vector. The one you present above is almost certainly a symmetrically redundant (equivalent) rotation to one within the fundamental region.

**EDIT:** see: https://doi.org/10.1016/S0045-7825(97)00072-8

**Reply by xshang93** [reply]
> @xshang93: You have to return your Rodrigues vector to the fundamental region for the crystal system considered. You will then not have such a large Rodrigues vector. The one you present above is almost certainly a symmetrically redundant (equivalent) rotation to one within the fundamental region.

Thank you @mkasemer . I think that makes sense. But if a rotation like this, i.e., not within the fundamental region, is supplied in neper, how does neper colour it? Does neper return it back to the fundamental space and calculate a colour, or does it just generate some random colour? I did a quick test:
1. Use a Rodrigues vector that is not within the fundamental space to generate an image
2. Get the RGB value of the generated grain with an colour extractor
3. Calculate the Rodrigues vector from that RGB value, with the formula from @rquey
4. Use the new Rodrigues vector to generate another image
The result gave me two different colours, but they are close (it is the top left grain that I'm playing with):
![26_4_3_4_geom_euler_ro](https://user-images.githubusercontent.com/86078353/168910356-957f05bf-f143-4e1c-966a-b149b81ce4af.png)
![26_4_3_4_geom_euler_rofromrgb](https://user-images.githubusercontent.com/86078353/168910357-3532319c-94b3-4401-8f1e-df7572dc4cbc.png)

**Reply by mkasemer** [reply]
@rquey would have to answer that specific question. When generating tessellations / orientations, do you specify the crystal symmetry?

**Reply by xshang93** [reply]
I'm using the default, i.e., triclinic when generating tessellations in neper. The orientations are generated in MTEX, where I first sample randomly from uniformly distributed orientations with a cubic symmetry (432), which is the orientations of the prior beta grains. Then I convert the orientations by Burger's relation to the orientations of the Alpha colonies, which has a Hexagonal (622) symmetry. 
I always generate the Euler angles wrt the global coordination system, and I thought the Euler angles should be independent from the crystal symmetry?
Am I doing anything wrong? Honestly I'm not sure what I'm doing is right. 

EDIT: Just wanted to add that the orientations I supplied to neper are in Euler angles, which were automatically converted to Rodrigues vectors in the resulting .tess files.

**EDIT2:** Looks like when generating images, equivalent orientations will be coloured same regardless if it is in the fundamental region. I made a mistake in my experiment above, and the colour should remain invariant when projecting to the fundamental region.

**Reply by rquey** [reply]
I agree with @mkasemer's assessment.

Rodrigues orientation coloring always uses the formula {R,G,B}=255 x (r_i + sqrt(2) - 1) / (2 * (sqrt(2) - 1)). Cubic crystal symmetry is assumed (which is right only if the material has cubic symmetry...): among all symmetrically-equivalent orientations, the one located inside the fundamental region is considered to apply the formula.

In your validation above, I think the difference between your two images comes from the transformation to integer values and back-transformation to Rodrigues vector, which must induce some error. Sometimes (but rarely), it can make an orientation "jump" from one side of the fundamental region to the opposite side.

**Reply by xshang93** [reply]
@rquey Thank you. Yes I did make an error in the first validation, but I have made it to work after correcting it.

I think "Cubic crystal symmetry is assumed (which is right only if the material has cubic symmetry...)" this is where I lack knowledge of. Does this mean I can't assign orientations with Rodrigues vectors to HCP phases in neper? Or is it that if it's not cubic crystal symmetry the colour is not calculated using this formula?
My understanding is that, regardless of the type of crystal symmetry, the Euler angles should always be the same if it's rotating in the same global coordinate, so I assigned Euler angles as orientation. Is this not the case? I'm having a hard time understanding why this is only right for cubic symmetry materials.

**Reply by mkasemer** [reply]
@xshang93 : it is not that the orientation is different or changing, it is that it is a redundant orientation, based on the symmetry of the crystal. Euler angles also fall prey to this.

Take, for example, something being rotated about its "x" axis (you can ultimately find Euler angles for this, or the Rodrigues vector). Even without considering crystal symmetry, you can plainly see that a rotation of 720deg is the same as a rotation of 360deg. The 720deg rotation is symmetrical equivalent to 360deg, and is thus redundant. Now consider if what we are rotating has cubic symmetry: the 360deg rotation is equivalent to a 180deg rotation, which is itself equivalent to a 90deg rotation. In each case, the crystal will look the same, and thus the larger rotations are symmetrically equivalent, or redundant.

**Reply by xshang93** [reply]
@mkasemer Yes, I think I don't have confusion in the redundant orientations, and just like your explanation, I understand that for a cubic symmetry 360, 180, and 90 are equivalent. And this, as @rquey explained, is being taken care of by neper through only visualizing the equivalents in the fundamental region.

I guess what confuses me is that, as @rquey pointed out

>  "Rodrigues orientation coloring always uses the formula {R,G,B}=255 x (r_i + sqrt(2) - 1) / (2 * (sqrt(2) - 1)). Cubic crystal symmetry is assumed (which is right only if the material has cubic symmetry...)"

I'm not sure what this means. Does this mean that, if I'm working with a material with non cubic crystal symmetry (take hcp Ti as an example), neper would by default think it as a cubic symmetry and then project it to the fundamental region, thus making it not right? 

e.g., if i have a grain at 360deg (let's just assume the orientation can be described by only one angle) for a hex symmetry, when inputting it to neper, neper would take it as in cubic symmetry and thus think it's equivalent to 90deg, but in fact it's not since in hex symmetry a grain oriented at 360deg is not equivalent to 90deg?

**Reply by mkasemer** [reply]
Apologies for my pedantic response. I believe you can check this simply by considering a few orientations:

Try ~50deg about the y axis, or r ~= (0, 0.446, 0). This orientation is outside of the fundamental region for a cubic material, but inside of the fundamental region for a hexagonal material. Make sure to specify the crystal type when generating the tessellation / orientation, and replace the orientations with this.

**Reply by xshang93** [reply]
Please no need to apologize. I'm indeed relatively new to this field.

Here's what I tried:
```
neper -T -n 1 -id 1 -ori 'rodrigues(0,0.446,0)' -oricrysym cubic -o cubic
neper -T -n 1 -id 1 -ori 'rodrigues(0,0.446,0)' -oricrysym hexagonal -o hex
```
The resulting .tess files are attached. Just as you described, for the cubic case, neper convert it automatically to fundamental space in the .tess file, while for the hex case, r remains unchanged since it's in the hex fundamental region.
[cubic.txt](https://github.com/neperfepx/neper/files/8720382/cubic.txt)
[hex.txt](https://github.com/neperfepx/neper/files/8720383/hex.txt)

But here comes the interesting part:
I tried to visualize both with these two lines
```
neper -V cubic.tess -datacellcol 'ori' -datacellcolscheme 'rodrigues' -print cubic
neper -V hex.tess -datacellcol 'ori' -datacellcolscheme 'rodrigues' -print hex
```
and I get the same result, which is exactly the colour one would expect for that r value with cubic symmetry:
![cubic](https://user-images.githubusercontent.com/86078353/169142595-e8b1b9cb-5ed9-4cbf-bdb5-b0548ce6cb00.png)
![hex](https://user-images.githubusercontent.com/86078353/169142608-0a9bbd46-ca1d-4fdb-a6fe-615d44513272.png)

So I think it's safe to conclude:
- when an orientation and a crystal symmetry is specified during tessellation, the orientation will be converted to the corresponding fundamental region in the resulting .tess file;
- during visualization, neper will take everything as if they have cubic symmetry and will convert that Rodrigues vector into cubic fundamental region and generate the RGB colour accordingly.

So this brings up the question for me: Since I'm visualizing dual phase Ti64, how would I do it via neper -V module if it doesn't plot orientations with hex symmetry? I recall that in your paper https://doi.org/10.1016/j.jmps.2017.03.013 the microstructure instances of Ti64 are properly generated, is there something I'm missing?

**Reply by rquey** [reply]
To summarize (I guess), whatever happens before, `rodrigues` orientation coloring will assume cubic crystal symmetry.

Obviously, this is not right, but Neper is sometimes biased toward cubic symmetry...

Proper coloring for hexagonals would require to define a proper color key for it; something like [this](https://neper.info/tutorials/orientation_color_key.html#rodrigues-color-key) but for the hexagonal fundamental region...

**Reply by xshang93** [reply]
Thanks, @rquey . That clarifies things. However, even with a proper colour key for the hex fundamental region, does Neper allow one to colour two different phases (with different crystal symmetries) in one microstructure?

**Reply by rquey** [reply]
When we have color keys for something else than cubic, it will!

**Reply by xshang93** [reply]
That would lead to my final (i guess) question, is there some software you know that could properly colour the microstructures I'm describing, and is available now? Thanks!

**Reply by rquey** [reply]
No, I don't, but perhaps others can suggest.

Of note, defining a color key in the hex fundamental region may not be so difficult, or other color keys might be useful...

Always happy to add new capabilities.

**Reply by xshang93** [reply]
So all that's missing is the colour key? Do you also need to work on specifying two different crystal symmetries during tessellation?

**Reply by mkasemer** [reply]
When I color HCP microstructures by orientation, I generally color by the angle between the c-axis of the crystal and the loading direction (or any sample direction). Since it exhibits elastic transverse isotropy, we only care about this angle for the directional modulus, and with few exceptions, primary slip family is determined singularly by this angle. However, this may be too simple overall...

**Reply by xshang93** [reply]
That sounds interesting. Basically what I want to achieve is to use colours to show the modulus along the loading direction. Since the modulus is a function of the orientations (for HCP), I hope to achieve this through the orientations. That sounds a lot like what you are describing. May I know what software you are using to do it, or it is some in-house developed scripts that you are using?

**Reply by mkasemer** [reply]
You can calculate the angle between the c-axis and any other vector (e.g., a sample axis / loading direction) readily with a rotation matrix / dot product between two vectors. From there, you can write this angle (calculated per-element) to a VTK file and visualize via Paraview. Neper may already be able to do this via mathematical manipulation of FEPX output (see `neper -S` section of the manual). Be careful to only consider angles 0-90, as anything less or greater is redundant / symmetrically equivalent.

**Reply by rquey** [reply]
Once the values are computed and recorded to a file, you can simply use `-V -dataelsetcol "real:file(...)"` (or `-dataeltcol ...` for per-element) to visualize.

If you have a simulation directory, you can add the data to it using [-S -reselset "\<res>:file(\<basename>)"](https://neper.info/doc/neper_s.html#cmdoption-res-cell-tess-node-elt-elset-mesh-entity) (you'll need all `<basename>.step#` files) and then simply use `-V -dataelsetcol real:<res>` to visualize.

**Reply by xshang93** [reply]
Amazing! Thanks @mkasemer and @rquey for the directions. I will test it out!

**Reply by xshang93** [reply]
Hi @mkasemer , sorry I just have an add-on question on FEPX: What we have been discussing here is only in concern with visualization right? I.e., when running simulations in FEPX, the crystal symmetries (and the corresponding orientations) will still be regulated as specified in the `.config` file, regardless how the crystal symmetry is specified in Neper? Moreover I guess redundant orientations will not cause any trouble to the simulation results, am I correct?

**Reply by mkasemer** [reply]
FEPX only cares about crystal orientation for elastic / plastic (slip) behavior. Orientations can be redundant - they will ultimately be converted to rotation matrices in-code.
